### ISSUE TITLE
TUnuranContDist ctor clones the dpdf but does not use the clone causing a segfault later on

### ISSUE THREAD

**Issue by AntoniMarcinek** [root-project/root#8196]
In the line 43 of TUnuranContDist.cxx there is
```
if (fDPdf) fDPdf->Clone();
```
which should be
```
if (fDPdf) fDPdf = fDPdf->Clone();
```

Otherwise the destructor of `TUnuranContDist` causes a segmentation violation if the ctor was called with non-zero dpdf.

**Comment by AntoniMarcinek** [comment]
Is there any way to signal that this is a bug, as it used to be in JIRA? Or a priority?

**Comment by jblomer** [comment]
@AntoniMarcinek Thanks for the report! I marked it as a bug via label assignment.

**Comment by AntoniMarcinek** [comment]
I would also be nice to have a bit more consistent interface. Right now pdf and cdf can be passed via references, while dpdf only as a pointer; pdf and dpdf via constructor, while cdf requires method call. Could you add the following:
```
TUnuranContDist(bool copyFunc=false)
SetPdf(const ROOT::Math::IGenFunction& pdf, bool isLogPdf=false)
SetDpdf(const ROOT::Math::IGenFunction& dpdf)
```
?

**Comment by lmoneta** [comment]
I will try to have a fix for this also for the next 6.24 release

**Comment by lmoneta** [comment]
> It would also be nice to have a bit more consistent interface. Right now pdf and cdf can be passed via references, while dpdf only as a pointer; pdf and dpdf via constructor, while cdf requires method call. Could you add the following:
> 
> ```
> TUnuranContDist(bool copyFunc=false)
> SetPdf(const ROOT::Math::IGenFunction& pdf, bool isLogPdf=false)
> SetDpdf(const ROOT::Math::IGenFunction& dpdf)
> ```
> 
> ?

The reason pdf is passed as a pointer and not reference is because it is an optional parameter, it is not required. 
I can add a SetDpdf() function, as (SetCDF), but I am not sure it is needed a SetPdf. One can use the constructor in this case.

**Comment by AntoniMarcinek** [comment]
Great regarding the 6.24.

As for the interface, there are Unuran methods (HINV and NINV) that do not require PDF, only CDF. This is why I proposed a constructor which only sets the `fOwnFunc` and let all 3  xxDF functions be optional. I see there is a ctor accepting TF1 pointers that can be used in this case, but it sets unconditionally the `fOwnFunc`. With that in mind, my proposed ctor has to be:
```cpp
TUnuranContDist(bool copyFunc);
```
i.e. without a default value for copyFunc. So for the case when the user wants only CDF and not to copy it in TUnuranContDist, a new ctor or a method to set `fOwnFunc` is needed anyway.

For the case where the user wants to set all 3 xxDF functions indeed the ctor is there, my point is just that the user code looks a bit ugly:
```cpp
TUnuranContDist dist(pdf, &dpdf, false, true);
dist.SetCdf(cdf);
```
compared to:
```cpp
TUnuranContDist dist(pdf, dpdf, cdf, false, true);
```
or:
```cpp
TUnuranContDist dist(true);
dist.SetPdf(pdf);
dist.SetDpdf(dpdf);
dist.SetCdf(cdf);
```

**Comment by lmoneta** [comment]
The current idea is the following: 

1. For algorithms using only pdf: 
```
TUnuranContDist dist(pdf); 
```
2. For algorithm using pdf + its derivative: 
```
TUnuranContDist dist(pdf, &dpdf); 
```
3.  For algorithms using pdf + derivative + pdf: 
```
 TUnuranContDist dist(pdf, &dpdf); 
dist.SetCdf(cdf);
```
4. For algorithms using only the cdf: 
```
TUnuranContDist dist(); 
dist.SetCdf(cdf);
```

I agree cases 3 and 4 are not very nice, also in case 4 one is forced to have the input functions cloned. 
I would prefer to not change the existing constructor for not breaking the backward compatibility. 
What about adding a constructor like this one: 
```
TUnuranContDist dist(const ROOT::Math::IGenFunction * pdf, 
                                       ROOT::Math::IGenFunction * dpdf, 
                                       ROOT::Math::GenFunction * cpdf,  bool isLogPdf = false, bool copyFunc = false); 
                                       ```
and similar for the TF1 case ? 

Thank you for the feedback !

**Comment by AntoniMarcinek** [comment]
I am not proposing to change a constructor, but rather add the one I proposed + the missing methods.

In general I rather prefer interfaces with references not pointers and limited number of parameters, but if for some reason you do not like the constructor + 2 methods I am proposing, then fine.

As for `TF1`, I have an impression, that the math libraries go into the direction of `ROOT::Math::` interfaces, so maybe there is no point in encouraging use of `TF1` for this kind of tasks (i.e. tasks going beyond drawing and `TH1::Fit`), but I understand the consistency argument in this case.

**Comment by lmoneta** [comment]
I have added 2 new constructor where you can pass both derivative of pdf and integral (cdf). See  #8254

**Comment by AntoniMarcinek** [comment]
Thanks!

**Comment by lmoneta** [comment]
This PR is changing the interface, so I don;t want to add as it is in 6.24. We should add just the fix for the segfault  in 6.24

**Comment by AntoniMarcinek** [comment]
Sure. Please let me know when the fix arrives in 6.24 patches. Btw. when do you expect new 6.24 patch release?
